File: GWT Chrome Extensions API/src/com/google/gwt/chrome/runtime/Runtime.java
package com.google.gwt.chrome.runtime;

import com.google.gwt.core.client.GWT;

public class Runtime implements ChromeRuntime {
	
	
	
	private final ChromeRuntime impl;
	public Runtime(){
		if(isApiAvailable()){
			impl = GWT.create(RuntimeImpl.class);
		} else {
			impl = GWT.create(RuntimeCsImpl.class);
		}
	}
	
	private final native boolean isApiAvailable() /*-{
		return !!(chrome.runtime);
	}-*/;

	@Override
	public void getId(RuntimeStringHandler handler) {
		impl.getId(handler);
	}

	@Override
	public void getManifest(ManifestHandler handler) {
		impl.getManifest(handler);
	}

	@Override
	public void getURL(String path, RuntimeStringHandler handler) {
		impl.getURL(path, handler);
	}

	@Override
	public void addOnInstalledHandler(RuntimeHandler handler) {
		impl.addOnInstalledHandler(handler);
	}

	@Override
	public void addOnSuspendHandler(RuntimeHandler handler) {
		impl.addOnSuspendHandler(handler);
	}

	@Override
	public void addOnSuspendCanceledHandler(RuntimeHandler handler) {
		impl.addOnSuspendCanceledHandler(handler);
	}

	@Override
	public void openOptionsPage(RuntimeHandler handler) {
		impl.openOptionsPage(handler);
	}

	@Override
	public void setUninstallURL(String url, RuntimeHandler handler) {
		impl.setUninstallURL(url, handler);
	}

	@Override
	public void reload() {
		impl.reload();
	}

	@Override
	public void requestUpdateCheck(RuntimeUpdateCheckHandler handler) {
		impl.requestUpdateCheck(handler);
	}

	/*@Override
	public void sendMessage(String extensionId, Object message, JavaScriptObject options,
			RuntimeJsonObjectHandler handler) {
		impl.sendMessage(extensionId, message, options, handler);
		
	}*/

	@Override
	public void getPlatformInfo(PlatformInfoHandler handler) {
		impl.getPlatformInfo(handler);
	}
}

File: gwt-chrome-api/DevExtension/apis/runtime.js
'use strict';
/*******************************************************************************
 * Copyright 2012 Pawel Psztyc
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/

/* globals chrome, gwt */

if (!chrome.runtime.getManifest) {
  let manifest = {
    version: 'dev-env',
    permissions: []
  };
  /**
   * Mock the getManifest function.
   */
  chrome.runtime.getManifest = function() {
    return manifest;
  };
  chrome.runtime.getManifestAsync = function(callback) {
    let p = {
      payload: 'runtime.getManifest',
      params: undefined
    };
    gwt.dev.chrome.addCallback(p.payload, p.params, callback);
    gwt.dev.chrome.postMessage(p);
  };
  chrome.runtime.getManifestAsync(function(result) {
    if (result) {
      manifest = result;
    }
  });
}
if (!chrome.runtime.openOptionsPage) {
  chrome.runtime.lastError = null;

  chrome.runtime.getId = function(callback) {
    let p = {
      payload: 'runtime.id',
      params: undefined
    };
    gwt.dev.chrome.addCallback(p.payload, p.params, callback);
    gwt.dev.chrome.postMessage(p);
  };
  chrome.runtime.getId((id) => chrome.runtime.id = id);

  let functions = '' +
  'chrome.runtime.openOptionsPage(callback);chrome.runtime.setUninstallURL(url,' +
  'callback);chrome.runtime.setUninstallURL(url,callback);chrome.runtime.reload();chrome.runtime' +
  '.requestUpdateCheck(callback);chrome.runtime.restart();chrome.runtime.sendMessage(extensionId' +
  ',message,options,callback);chrome.runtime.sendNativeMessage(application,message,callback);' +
  'chrome.runtime.getPlatformInfo(callback);chrome.runtime.getPackageDirectoryEntry(callback);' +
  'chrome.runtime.onStartup.addListener(callback);chrome.runtime.onInstalled.addListener(' +
  'callback);chrome.runtime.onSuspend.addListener(callback);chrome.runtime.onSuspendCanceled.' +
  'addListener(callback);chrome.runtime.onUpdateAvailable.addListener(callback);chrome.runtime.' +
  'onConnect.addListener(callback);chrome.runtime.onConnectExternal.addListener(callback);chrome' +
  '.runtime.onMessage.addListener(callback)';

  gwt.dev.chrome.registerAPI(functions.split(';'));
}
